### PR TITLE
[core] Add ray.util.list_actors() API

### LINKED ISSUE #16623: [RFC] API to list named actors and placement groups
In the Ray Serve controller, we currently need to checkpoint the names of all of the actors and placement groups in the system to avoid leaking actors in failure scenarios. This logic could be simplified if we had the ability to query all named actors and placement groups in the system (or a namespace) instead. This would allow Serve to recover the current state of the system by querying the GCS and only checkpoint the goal state, rather than attempting to maintain checkpoints for both the current and goal states.

Proposed API:
```python
ray.list_named_actors() -> List[str]
ray.list_named_placement_groups() -> List[str]
```

This would return a list of the currently active actor and PG names in the current namespace. I'm not too opinionated on the exact API naming here. The implementation for this should be very simple - just a simple RPC to the GCS.

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Introduces a new top-level API to get a list of the named actors currently active in the system: `ray.util.list_named_actors()`. This returns a list of all named actors that are currently managed by the GCS, regardless of their stage in the lifecycle (pending, running, or being restarted).

By default, the list is scoped to the current driver's namespace, but an optional flag `all_namespaces` can be passed to return all actors in the cluster.

The format of the returned list is:

```python
["name1", "name2", ...]
```

If `all_namespaces` is passed, then the names are prefixed with their namespace:
```python
["namespace1/name1", "namespace2/name2"]
```

## Related issue number

Closes https://github.com/ray-project/ray/issues/16623

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/protobuf/gcs_service.proto [hunk 1]
```diff
@@ -98,6 +98,16 @@ message GetNamedActorInfoReply {
   ActorTableData actor_table_data = 2;
 }
 
+message GetAllActorNamesRequest {
+  // Namespace to filter to.
+  string ray_namespace = 2;
```
Comment (ericl) [inline src/ray/protobuf/gcs_service.proto:103]: Could we say this is empty for all namespaces?

File: src/ray/protobuf/gcs_service.proto [hunk 2]
```diff
@@ -98,6 +98,16 @@ message GetNamedActorInfoReply {
...
+}
+
+message GetAllActorNamesReply {
+  GcsStatus status = 1;
+  repeated string actor_name_list = 2;
```
Comment (ericl) [inline src/ray/protobuf/gcs_service.proto:108]: How about `repeated Actor` and

```
message Actor {
  string name;
  string namespace;
}
```

File: src/ray/protobuf/gcs_service.proto [hunk 3]
```diff
@@ -98,6 +98,16 @@ message GetNamedActorInfoReply {
   ActorTableData actor_table_data = 2;
 }
 
+message GetAllActorNamesRequest {
+  // Namespace to filter to.
+  string ray_namespace = 2;
+}
+
+message GetAllActorNamesReply {
```
Comment (ericl) [inline src/ray/protobuf/gcs_service.proto:106]:
```suggestion
message ListActorsReply {
```

**Review by ericl — COMMENTED** [review]
Couple questions:
- why not extend the GetAllActorData RPC?
- why get_() instead of list_()?

File: src/ray/protobuf/gcs_service.proto (hunk 1 above)
Comment (WangTaoTheTonic) [inline src/ray/protobuf/gcs_service.proto:103]: Why starts with 2 instead of 1? :-p

File: src/ray/protobuf/gcs_service.proto (hunk 3 above)
Comment (WangTaoTheTonic) [reply src/ray/protobuf/gcs_service.proto:106]: I don't understand this suggested change. :(

File: python/ray/includes/libcoreworker.pxd [hunk 4]
```diff
@@ -161,6 +161,7 @@ cdef extern from "ray/core_worker/core_worker.h" nogil:
         ActorHandleSharedPtr GetActorHandle(const CActorID &actor_id) const
         pair[ActorHandleSharedPtr, CRayStatus] GetNamedActorHandle(
             const c_string &name)
+        CRayStatus GetAllActorNames(c_vector[c_string] *actor_names)
```
Comment (WangTaoTheTonic) [inline python/ray/includes/libcoreworker.pxd:164]: All other `Getxxxx` requests return values in `return` type instead of `reference parameter`, can we make this consistent with them?

**Comment by edoakes** [comment]
@ericl will address the smaller comments when I'm back from OOO.

- `get_xxx` sounds fine to me (matches existing APIs better anyways)
- The `GetAllActorData` RPC is doing something pretty different - it returns a dump of the `ActorTable` data which is info about the "physical" actors in the system (actual actor processes that are pending/running/exited). Here we just care about the "logical" actors (those that the GCS currently has in scope). Concretely, AFAICT the former would not always work in the case where an actor has crashed and will be restarted (but hasn't been yet). I could piggy back this logic onto the same RPC but it would basically be two separate handlers/implementations with a flag to toggle between them so I figured it's better to keep them orthogonal. If you'd strongly prefer that I can refactor but IMO it's best to just this simple call simple.

File: src/ray/protobuf/gcs_service.proto (hunk 1 above)
Comment (edoakes) [reply src/ray/protobuf/gcs_service.proto:103]: Unfortunately an empty namespace is valid...

File: src/ray/protobuf/gcs_service.proto (hunk 2 above)
Comment (edoakes) [reply src/ray/protobuf/gcs_service.proto:108]: I don't currently see a use case for getting an actor name that isn't in the current namespace so I'd prefer to just scope this to the minimal version we need for now

File: python/ray/includes/libcoreworker.pxd (hunk 4 above)
Comment (edoakes) [reply python/ray/includes/libcoreworker.pxd:164]: done

File: src/ray/core_worker/core_worker.cc [hunk 5]
```diff
@@ -1715,7 +1715,7 @@ Status CoreWorker::CreateActor(const RayFunction &function,
   TaskSpecification task_spec = builder.Build();
   Status status;
   if (options_.is_local_mode) {
-    if (task_spec.IsDetachedActor()) {
+    if (!actor_name.empty()) {
```
Comment (edoakes) [inline src/ray/core_worker/core_worker.cc:1718]: small drive-by bug fix for local mode

File: src/ray/protobuf/gcs_service.proto [hunk 6]
```diff
@@ -135,6 +145,8 @@ service ActorInfoGcsService {
   rpc GetActorInfo(GetActorInfoRequest) returns (GetActorInfoReply);
   // Get actor data from GCS Service by name.
   rpc GetNamedActorInfo(GetNamedActorInfoRequest) returns (GetNamedActorInfoReply);
+  // Get all named actor names from GCS Service.
+  rpc GetAllActorNames(GetAllActorNamesRequest) returns (GetAllActorNamesReply);
   // Get information of all actor from GCS Service.
   rpc GetAllActorInfo(GetAllActorInfoRequest) returns (GetAllActorInfoReply);
```
[the remark this answers is not in the document]
Comment (edoakes) [reply src/ray/protobuf/gcs_service.proto:151]: see comment in thread

**Comment by edoakes** [comment]
@ericl can you take another look here? Small comments addressed.

- I think we should stick with `get_actor_names` because it matches existing APIs (not that opinionated though and waiting for @pcmoritz to chime in).
- I'd prefer to leave this orthogonal to the existing `GetActorInfo` API which is effectively an "actor table dump."

File: src/ray/protobuf/gcs_service.proto (hunk 2 above)
Comment (ericl) [reply src/ray/protobuf/gcs_service.proto:108]: Isn't it important to list actors across namespaces? I don't think adding namespace changes more than a few lines of code and makes this API much nicer.

**Comment by ericl** [comment]
Can we rename it to list_actors()? Getting names is oddly narrow of an API.

File: src/ray/protobuf/gcs_service.proto (hunk 2 above)
Comment (edoakes) [reply src/ray/protobuf/gcs_service.proto:108]: What's the use case for listing actors across namespaces? It's sort of the whole point of namespaces to isolate these

**Comment by ericl** [comment]
How do I find all named actors in a cluster for debugging, or if I lost my
anonymous namespace?

I think the API could be much more generally useful with just this small
tweak.

File: src/ray/protobuf/gcs_service.proto (hunk 2 above)
Comment (robertnishihara) [reply src/ray/protobuf/gcs_service.proto:108]: >What's the use case for listing actors across namespaces? It's sort of the whole point of namespaces to isolate these

I have the same question as @edoakes.

Comment (ericl) [reply src/ray/protobuf/gcs_service.proto:108]: How do I find all named actors in a cluster for debugging, or if I lost my anonymous namespace?

I think the API could be much more generally useful with just this small tweak.

**Comment by edoakes** [comment]
@ericl updated, PTAL

**Comment by edoakes** [comment]
Ping again @pcmoritz

**Review by wuisawesome — APPROVED** [review]

**Comment by AmeerHajAli** [comment]
@ijrsvt , can you please review this and make sure it has everything necessary for Ray Client?

**Comment by edoakes** [comment]
Ah yeah good call @AmeerHajAli, a pointer on the best way to add tests would be great as well.

**Review by ericl — APPROVED** [review]

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 7]
```diff
@@ -200,6 +200,21 @@ Status ServiceBasedActorInfoAccessor::AsyncGetByName(
   return Status::OK();
 }
 
+Status ServiceBasedActorInfoAccessor::AsyncListActors(
+    bool all_namespaces, const std::string &ray_namespace,
+    const ItemCallback<std::vector<rpc::NamedActorInfo>> &callback) {
+  RAY_LOG(DEBUG) << "Getting all actor names";
```
Comment (WangTaoTheTonic) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:206]: The log remains same with old api name.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 8]
```diff
@@ -462,6 +480,26 @@ ActorID GcsActorManager::GetActorIDByName(const std::string &name,
   return actor_id;
 }
 
+std::vector<std::pair<std::string, std::string>> GcsActorManager::ListActors(
+    bool all_namespaces, const std::string &ray_namespace) const {
+  std::vector<std::pair<std::string, std::string>> actors;
+  if (all_namespaces) {
+    for (auto namespace_it : named_actors_) {
```
Comment (WangTaoTheTonic) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:487]: add `const &` in those for loops?

**Review by WangTaoTheTonic — COMMENTED** [review]
I agree that the API should be more general, but should we at least change the api to List`Named`Actor?

**Comment by ijrsvt** [comment]
@edoakes you will need to add this to  `python/ray/util/client/api.py::ClientAPI`. This will require implementing a gRPC service for `ray_client.proto::ClusterInfoType::TypeEnum`.

**Comment by robertnishihara** [comment]
@edoakes @pcmoritz I think it'd be much more clear to say `[get or list]_named_actors` since people who are using regular actors will run this and wonder why nothing is returned.

Also, maybe stick it in `util` or `experimental` to avoid cluttering up the top level namespace?

**Comment by edoakes** [comment]
I'm fine with list_named_actors, though it's less extensible in the future. I'll defer to @pcmoritz's judgement.

**Comment by pcmoritz** [comment]
Ok I don't think we have a full picture of what exactly we want here, so how about we do `ray.util.list_named_actors()` which is the least confusing option for users for now (that might be all we need going forward -- maybe we want to extend this to all actors in the future and maybe we don't; if we do we should re-design this and do `ray.list_actors()` and can keep this API working).

For the actual semantics I think we should do this:

`ray.util.list_named_actors()` returns a list ["actor1", "actor2", "my_other_actor"] of all named actors in the current namespace.

`ray.list_named_actors(all_namespaces=True)` returns a list ["namespace1/actor1", "namespace/actor2", "namespace1/my_other_actor", "namespace2/yet_another_actor"]` of all named actors in all namespaces.

At the same time we should disallow "/" from namespaces and actor names. In the future you might be able to access actors via a URL like `ray://12.78.45.3:6379/namespace1/actor1` if you have the necessary credentials to access it.

The rational for this:
- Make it play well with `ray.get_actor` and allow to pass in results from `list_named_actors` into `ray.get_actor`.
- In this case by default Ray users don't need to know about namespaces if they don't need them, which is more beginner friendly.

**Comment by edoakes** [comment]
@pcmoritz description and PR have been updated to reflect your suggestion.
